### PR TITLE
SQL: Make CASTing string to DATETIME more lenient

### PR DESCRIPTION
**Pull request by matriv**
Some BI tools (i.e. Tableau) would try to cast strings where the time
part is separated from the date part with a whitespace instead of `T`.
Adjust type conversion used by CAST to support this.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java [hunk 1]
```diff
@@ -33,10 +54,21 @@ public static ZonedDateTime asDateTime(long millis) {
      * Parses the given string into a DateTime using UTC as a default timezone.
      */
     public static ZonedDateTime asDateTime(String dateFormat) {
-        return DateFormatters.from(UTC_DATE_TIME_FORMATTER.parse(dateFormat)).withZoneSameInstant(UTC);
+        int separatorIdx = dateFormat.indexOf('-');
+        if (separatorIdx == 0) { // negative year
+            separatorIdx = dateFormat.indexOf('-', 1);
+        }
+        separatorIdx = dateFormat.indexOf('-', separatorIdx + 1) + 3;
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:61]: Why '+3' here?

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java [hunk 2]
```diff
@@ -6,19 +6,40 @@
...
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_TIME;
 
 //NB: Taken from sql-proto.
 public final class DateUtils {
 
     public static final ZoneId UTC = ZoneId.of("Z");
 
-    private static final DateFormatter UTC_DATE_TIME_FORMATTER = DateFormatter.forPattern("date_optional_time").withZone(UTC);
+    private static final DateTimeFormatter DATE_OPTIONAL_TIME_FORMATTER_WHITESPACE = new DateTimeFormatterBuilder()
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:25]: Why are these needed in QL and not SQL only?
Also, SQL has a `DateUtils` class with DateTimeFormatters that use whitespace separators. Haven't looked into this in more depth, but have you considered having a single definition source for these DateTimeFormatters?

**Review by astefan — COMMENTED** [review]
Left two questions, mostly investigative, not necessarily suggesting to make any changes. Just making sure all facets were covered.
Also, no integration testing? :-)

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:61]: if you have a date like `-1000-11-22T10:20:30....`:
check for the first char, if `-` move to idx 1.
Then detect the first `-`, then detect the second `-` and `+3` to move past `22` to the `T` or whitespace.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java (hunk 2 above)
Comment (matriv) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:25]: Currently SQL is using the QL for the String->DATETIME conversion, and SQL has the additional conversions for String->DATE, String->TIME. If we move the leniency only to SQL we'll have to move that conversion there. What do you think?

Also we have this issue for consolidating the date related utils: #56494 so I'd rather do the whole refactoring in a separate PR.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:61]: Worth a comment in the code.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java (hunk 2 above)
Comment (costin) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DateUtils.java:25]: It's annoying the format builder does not allow for two literals to be specified as either/or yet supports optional.

**Review by costin — APPROVED** [review]
LGTM

**Review by astefan — APPROVED** [review]
LGTM

**Review by bpintea — COMMENTED** [review]
LGTM
